swarm/src/handler/pending.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2022 Protocol Labs.
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::handler::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    SubstreamProtocol,
+};
+use crate::NegotiatedSubstream;
+use libp2p_core::{
+    upgrade::{InboundUpgrade, OutboundUpgrade, PendingUpgrade},
+    Multiaddr,
+};
+use std::task::{Context, Poll};
+use void::Void;
+
+/// Implementation of [`ConnectionHandler`] that returns a pending upgrade.
+#[derive(Clone, Debug)]
+pub struct PendingConnectionHandler {
+    protocol_name: String,
+}
+
+impl PendingConnectionHandler {
+    pub fn new(protocol_name: String) -> Self {
+        PendingConnectionHandler { protocol_name }
+    }
+}
+
+impl ConnectionHandler for PendingConnectionHandler {
+    type InEvent = Void;
+    type OutEvent = Void;
+    type Error = Void;
+    type InboundProtocol = PendingUpgrade<String>;
+    type OutboundProtocol = PendingUpgrade<String>;
+    type OutboundOpenInfo = Void;
+    type InboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(PendingUpgrade::new(self.protocol_name.clone()), ())
+    }
+
+    fn inject_fully_negotiated_inbound(
+        &mut self,
+        _: <Self::InboundProtocol as InboundUpgrade<NegotiatedSubstream>>::Output,
+        _: Self::InboundOpenInfo,
+    ) {
+        unreachable!("`PendingUpgrade` is never successful.");
```

Comment:
Why is this different from `inject_fully_negotiated_outbound`?

Reply:
In `inject_fully_negotiated_outbound` we can leverage the `OutboundOpenInfo` to carry an uninhabited type (in this case `Void`), thus proving it to be impossible at compile time. for `inject_fully_negotiated_inbound` we don't have a corresponding `InboundOpenInfo` mechanism, and can thus not apply this trick.

Reply:
But we do or what else is line 62? 🤔

Reply:
> for `inject_fully_negotiated_inbound` we don't have a corresponding `InboundOpenInfo` mechanism, and can thus not apply this trick.

Sorry. This is wrong. We do have the `InboundOpenInfo` mechanism. The problem is that in `ConnectionHandler::listen_protocol` and within that the call to `SubstreamProtocol::new` we have to provide the `InboundOpenInfo`. We can not provide an instance of `Void`, thus we can not use `Void` and `InboundOpenInfo` to make this state impossible at compile time.

@thomaseizinger does that make sense?

Reply:
Ah yes that makes perfect sense!

Couldn't we use the actual substream variable though? The upgrade uses pending so we don't have to instantiate `void` but it would be "resolved" here, correct?

Reply:
Good point. Thanks @thomaseizinger. What do you think of https://github.com/libp2p/rust-libp2p/pull/2785/commits/a0960bb091b8d923c05baa298b241c4e2dc08aeb?